File: cses/04_graph_algorithms/1676_road-construction/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    # DSU: parent lưu cha của mỗi đỉnh, size lưu kích thước mỗi tập.
    parent = list(range(n + 1))
    size = [1] * (n + 1)

    components = n              # Số thành phần liên thông hiện tại.
    max_size = 1 if n >= 1 else 0  # Kích thước thành phần lớn nhất.

    def find(x):
        # Tìm root của tập chứa x, nén đường đi kiểu lặp (tránh đệ quy sâu).
        root = x
        while parent[root] != root:
            root = parent[root]
        while parent[x] != root:
            parent[x], x = root, parent[x]
        return root

    out = []
    for _ in range(m):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        ra = find(a)
        rb = find(b)
        # Chỉ gộp khi a và b thuộc hai thành phần khác nhau.
        if ra != rb:
            # union by size: gắn cây nhỏ vào cây lớn.
            if size[ra] < size[rb]:
                ra, rb = rb, ra
            parent[rb] = ra
            size[ra] += size[rb]
            components -= 1
            if size[ra] > max_size:
                max_size = size[ra]
        # Ghi lại kết quả sau mỗi con đường được xây.
        out.append(f"{components} {max_size}")

    sys.stdout.write("\n".join(out))
    if out:
        sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: cses/04_graph_algorithms/1676_road-construction/solution.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <utility>
#include <vector>

// Cấu trúc DSU (Union-Find) với path compression và union by size.
class DisjointSet {
public:
    explicit DisjointSet(int n) : parent_(n + 1), size_(n + 1, 1) {
        // Ban đầu mỗi đỉnh là cha của chính nó (mỗi tập chỉ có 1 phần tử).
        std::iota(parent_.begin(), parent_.end(), 0);
    }

    // Tìm đại diện (root) của tập chứa node, kèm nén đường đi.
    int find(int node) {
        int root = node;
        while (parent_[root] != root) {
            root = parent_[root];
        }
        while (parent_[node] != node) {
            const int next = parent_[node];
            parent_[node] = root;
            node = next;
        }
        return root;
    }

    // Gộp hai tập; trả về kích thước tập sau khi gộp (union by size).
    int unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) {
            return size_[a];
        }
        // Gắn cây nhỏ vào cây lớn để giữ cây thấp.
        if (size_[a] < size_[b]) {
            std::swap(a, b);
        }
        parent_[b] = a;
        size_[a] += size_[b];
        return size_[a];
    }

private:
    std::vector<int> parent_;
    std::vector<int> size_;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    DisjointSet dsu(n);
    int components = n;         // Số thành phần liên thông hiện tại.
    int largest_component = 1;  // Kích thước thành phần lớn nhất.
    for (int i = 0; i < m; ++i) {
        int a, b;
        std::cin >> a >> b;
        // Chỉ gộp khi hai thành phố thuộc hai thành phần khác nhau.
        if (dsu.find(a) != dsu.find(b)) {
            largest_component = std::max(largest_component, dsu.unite(a, b));
            --components;
        }
        // In kết quả sau mỗi con đường được xây.
        std::cout << components << ' ' << largest_component << '\n';
    }
    return 0;
}
